torch/_higher_order_ops/hints_wrap.py

```diff
@@ -0,0 +1,104 @@
+# mypy: allow-untyped-defs
+import copy
+
+import torch.utils._pytree as pytree
+from torch._C import DispatchKey
+from torch._higher_order_ops.utils import (
+    autograd_not_implemented,
+    reenter_make_fx,
+    unique_graph_id,
+)
+from torch._ops import HigherOrderOperator
+from torch._subclasses.fake_tensor import FakeTensorMode
+from torch.fx.experimental.proxy_tensor import ProxyTorchDispatchMode, track_tensor_tree
+
+
+# used for wrapping a function/op with context hints
+class HintsWrapper(HigherOrderOperator):
+    def __init__(self):
+        super().__init__("hints_wrapper")
+
+
+hints_wrapper = HintsWrapper()
+# Override hints_wrapper.__module__ to "torch.ops.higher_order" so that in the generated
+# graph module, hints_wrapper node's target is correctedly printed as torch.ops.higher_order.hints_wrapper
+hints_wrapper.__module__ = "torch.ops.higher_order"
+
+
+class no_hints_kwargs:
+    def __init__(self, kwargs):
+        self.kwargs = copy.copy(kwargs)
+
+    def __enter__(self):
+        if "hints" in self.kwargs:
+            del self.kwargs["hints"]
+        return self.kwargs
+
+    def __exit__(self, *args):
+        pass
+
+
+@hints_wrapper.py_impl(DispatchKey.CompositeExplicitAutograd)
+def hints_wrapper_dense(body_fn, *args, **kwargs):
```

Comment:
I'm thinking that we can explicitly change the signature of hints_wrapper to hints_wrapper(body_fn, args, kwargs, hints):

Concretely, for the example in test, the UX should look like
```
def fn_with_hints(x, y):
    x = x + y
    
    # we don't want inner_body_fn to take an un-used `hints` argument
    def inner_body_fn(x, y):
        x = torch.relu(x)
        x = x + y
        return x

    # we don't want outer_body_fn to take an un-used `hints` argument
    def outer_body_fn(x, y):
        x = hints_wrapper(inner_body_fn, x, y, hints={"inner_body": True})
        x = torch.abs(x)
        return x
    
    # We need to avoid naming collision with body_fn by
    # asking users explicitly provide args and kwargs:
    res = hints_wrapper(outer_body_fn, (x, y), {}, hints={"outer_body": True})

    # We could even provide a user-facing wrapper function e.g. with_hints, that calls into hints_wrapper to make it look nicer
    res = with_hints(outer_body_fn, hints={"outer_body": True})(x, y)
    return res
```
In the dispatch key implementation of hints_wrapper, we could expliclity list out the args and kwargs of body_fn and don't pass in the hints to body_fn since they're not needed:
```
@hints_wrapper.py_impl(DispatchKey.CompositeExplicitAutograd)
def hints_wrapper_dense(body_fn, args, kwargs, hints):
   return body_fn(*args, **kwargs)
```
We also need to change the dynamo implementation a bit to incorporate this signature change.

Reply:
@ydwu4  Thanks for advice. By below usage, do we need to preserve kwargs since it's an empty dict?

> res = hints_wrapper(outer_body_fn, (x, y), {}, hints={"outer_body": True})

Reply:
I think it would be easier if we always pass the kwargs in even if it's empty.

Reply:
@ydwu4 do you want this to be exportable? Export can't take dicts as operator inputs

Reply:
Do you mean serialization and deserialization part of export cannot support dict input?

Reply:
Just skip those serde tests for hints_wrapper.